### DISCUSSION TITLE
Strange Behavior with Union

### DISCUSSION THREAD

**Discussion by deweyjose** [maticzav/graphql-shield#1064] [Q&A]
Hi,

I'm seeing a strange issue where all data in my Union types is removed from the response when I used a shield. Even if the query isn't part of any rule.

Have you seen anything like this?

**Comment by maticzav** [comment]
Hi,

Please add a reproduction or a bit more than just "strange issue". I haven't seen anything like it so far. From the top of my head, however, I think this might stem from you not "allowing" the union type or one of the types it consists of.

I can probably help you better if you post permissions object.

**Comment by deweyjose** [comment]
Apologies on the vagueness ...

`gql`
schema {
  query: RootQueryType
}

type BookType {
  author: String!
}

type MovieType {
  director: String!
}

type RootQueryType {
  unionTest: SearchableType!
}

union SearchableType = MovieType | BookType
``



`shield(
  {
    BookType: allow,
    MovieType: allow,
    SearchableType: allow
  }
);`

**Reply by maticzav** [reply]
Could you merge the two comments and format the response? I need more information about what you've tried and what you think might be the problem. GraphQLShield doesn't manipulate the response, though. If it were a permissions issue, you wouldn't get any response in `unionTest` sine it's a non-nullable field.

Have you tried setting the debug option to true in shield and checking for any errors?

**Comment by deweyjose** [comment]
The unionTest query is hard coded to return `{ director: 'John' }`

This is what I'm seeing:

query test {
  unionTest {
    __typename 
    ... on MovieType {
      director
    }
    ... on BookType {
      author
    }
  }
}

{
  "data": {
    "unionTest": {
      "__typename": "MovieType"
    }
  }
}

**Comment by deweyjose** [comment]
@maticzav we figured it out on our end. 

The following code snippet will produce expected results now. I changed the return value in resolveType to a string of the class name rather than the value MovieType or BookType. One thing to note though, either return values for resolveType work if I do not applyMiddleware()  the schema. 

`
const MovieType = new GraphQLObjectType({
  name: 'MovieType',
  fields: {
    director: { type: new GraphQLNonNull(GraphQLString) }
  }
});

const BookType = new GraphQLObjectType({
  name: 'BookType',
  fields: {
    author: { type: new GraphQLNonNull(GraphQLString) }
  }
});

const DATA = [{ director: 'catherine hardwicke' }, { author: 'catherine cookson' }];

const SearchableType = new GraphQLUnionType({
  name: 'SearchableType',
  types: [MovieType, BookType],
  resolveType(data) {
    if (data.director) {
      return 'MovieType';
    }
    if (data.author) {
      return 'BookType';
    }
    return null;
  }
});
`
